__tests__/authenticationFlow.test.js

```diff
@@ -0,0 +1,324 @@
+    it('can sign up as new user', () => {
+      const env = createEnv()
+      env.signUp('DJTHAI', 'thai@bemuse.ninja', 'strongpassword').mustSucceed()
+      env.loginByUsernamePassword('DJTHAI', 'strongpassword').mustSucceed()
+      env.loginByUsernamePassword('thai@bemuse.ninja', 'strongpassword').mustSucceed()
+      return env.verify()
+    })
+    it('cannot sign up if legacy user already exist', () => {
+      const env = createEnv()
+      env.givenLegacyUser('DJTHAI', 'thai@another.place', 'wow')
+      env.signUp('DJTHAI', 'thai@bemuse.ninja', 'strongpassword').mustFail()
+      return env.verify()
+    })
+    it('cannot sign up if email already exist as legacy user', () => {
+      const env = createEnv()
+      env.givenLegacyUser('THAI', 'thai@bemuse.ninja', 'wow')
+      env.givenPlayer('THAI')
+      env.signUp('DJTHAI', 'thai@bemuse.ninja', 'strongpassword').mustFail()
+      return env.verify()
+    })
+    it('should link the player after successful', () => {
+      const env = createEnv()
+      env.signUp('DJTHAI', 'thai@bemuse.ninja', 'strongpassword').mustSucceed()
+      env.playerWithName('DJTHAI').shouldBeLinkedTo('thai@bemuse.ninja')
+      return env.verify()
+    })
+  })
+
+  describe('edge cases', () => {
+    describe('hijacking user by creating account with existing player id as username', () => {
+      it('should be prevented', () => {
+        const env = createEnv()
+        env.givenLegacyUser('THAI', 'thai@bemuse.ninja', 'wow')
+        env.givenPlayer('THAI')
+        env.externalSignUp('player0', 'player0@bemuse.ninja', 'strongpassword').mustFail()
+        return env.verify()
+      })
+    })
+    describe('an unlinked account (somehow)', () => {
+      it('should try to link with legacy user via email, if possible')
+      it('should ask for player name and try to register')
+    })
+  })
+})
+
+function createEnv () {
+  const legacyUserByUsername = { }
+  const legacyUserByEmail = { }
+  const playerByPlayerName = { }
+  const playerById = { }
+  const actions = [ ]
+  let nextPlayerId = 0
+
+  function queue (action) {
+    actions.push(action)
+  }
+
+  // This is Auth0.
+  const externalProvider = createExternalAuthProvider({
+    databaseLogIn (username, password) {
+      const user = legacyUserByEmail[username] || (() => {
+        const player = playerById[username]
+        return player && legacyUserByUsername[player.playerName]
+      })()
+      const player = playerByPlayerName[user.username]
+      if (user.password === password) {
+        return { username: player._id, email: user.email }
+      }
+      return null
+    },
+    databaseGetUser (username) {
+      const user = legacyUserByEmail[username] || (() => {
+        const player = playerById[username]
+        return player && legacyUserByUsername[player.playerName]
+      })()
+      return user
+    },
+    rule (user) {
+      if (user.email && !user.playerName) {
+        const player = link(user._id, user.username)
```

Comment:
How about linking at the client side?

This allows user-initiated linking (such as from Twitter) at the client-side.